### PR TITLE
Allow restricting cultures creation with any arbitrary names in Globalization Invariant Mode

### PR DESCRIPTION
**Pull request by tarekgh**
This change addresses the second part of #43774

This change is forcing to throw exception when creating any culture other than Invariant culture when enabling the Globalization Invariant Mode. This will allow easier diagnosis of the failures when turning on the Globalization Invariant Mode. 
When turning on the [Invariant globalization mode](https://docs.microsoft.com/en-us/dotnet/core/run-time-config/globalization?ranMID=46131&ranEAID=a1LgFw09t88&ranSiteID=a1LgFw09t88-TL.gIntYvAARbx7QNuWPjg&epi=a1LgFw09t88-TL.gIntYvAARbx7QNuWPjg&irgwc=1&OCID=AID2000142_aff_7806_1243925&tduid=%28ir__mrdukhopi0kfqxeykk0sohzifv2xuvvxe2xvqk3y00%29%287806%29%281243925%29%28a1LgFw09t88-TL.gIntYvAARbx7QNuWPjg%29%28%29&irclickid=_mrdukhopi0kfqxeykk0sohzifv2xuvvxe2xvqk3y00#invariant-mode) the config switch `System.Globalization.PredefinedCulturesOnly` will be automatically turned on. If anyone for any reason want to get back to the old behavior which is allowing creating any cultures by providing Invariant culture data, can do that by explicitly turning off the `System.Globalization.PredefinedCulturesOnly` switch.

The breaking change doc is tracked by [Restrict Cultures Creation in Globalization Invariant Mode](https://github.com/dotnet/docs/issues/24849) issue.

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @safern

File: src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs
Comment (safern) [inline src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs:157]: nit: typo, `invarintEnabled` => `invariantEnabled`.

**Review by safern — APPROVED** [review]
LGTM. It seems like there are some Globalization trimming tests failing.

File: src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs
Comment (eerhardt) [inline src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs:161]: `regaldless` => `regardless`

File: src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs
Comment (eerhardt) [inline src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs:165]: Should we test a couple basic operations work with invariant mode here? string operations, calendar, etc? Just super basic things

File: src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:16]: It looks like the trimming tests are failing because this new addition is causing the `Settings` class to be preserved when trimming with `Invariant=true`. This will also cause `AppContextConfigHelper` to be preserved in that mode.

@marek-safar - are we certain this new setting will provide enough value to justify adding it?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:16]: That was my point I raised in the comment https://github.com/dotnet/runtime/issues/43774#issuecomment-715627250. @marek-safar can advise if he still seeing a good value to have this or just not worth it.

File: src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs
Comment (tarekgh) [reply src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs:165]: I added a couple of more cases to test invariant.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: CC @jkotas for awareness

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: I don't see AllowInvariantCultureOnly setting to be ever used so it'd be better to have it always trimmed.

Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: If you don't see it ever being used, why are we adding it?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: @marek-safar please advise if this feature you originally requested is needed and will be consumed by some platforms/containers. If you think it is not clear where is going to be used, then I would hold on it till it will be clear it is really needed.

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: I think the setting would make more sense as opt-out mode. Right now I don't see we will advise developers which invariant mode they should be using especially if this one is not fully integrated (no msbuild property).

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: I agree with @marek-safar . I think this should be enabled by default for invariant mode. I think it is better to throw exceptions than to return bogus results (current behavior). If somebody is happy with the bogus results, they can opt-in into the somewhat broken behavior.

I understand that it means that some libraries will be broken by default in invariant mode. I think that it is a good thing. It will contribute towards being able to use the invariant mode with confidence.

Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: Are we concerned that this can break existing apps using Invariant mode? For example apps running on alpine?

cc @richlander @MichaelSimons

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: Yes, the change would have to be tracked as a .NET 6 breaking change. The apps that got broken and are happy with the bogus globalization behavior can opt-in into it.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: I am not agreeing with that. what benefit we are getting here more than creating burden on us asking every app running into this problem to enable this switch? why returning invariant data for any locale is bad or wrong?

File: src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs:33]: This can be combined with the `IsWindows()` check below for `is ConstructorInfo`.

Also - no need to check the name, this code is only look at static members, and if the member is a ConstructorInfo, it must be a static constructor.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs]: @jkotas @marek-safar my last commit implemented throwing by default when creating any culture different than Invariant culture when enabling the Globalization Invariant Mode. also, I am going to create a breaking change issue and link it to this PR too. Could you please have a look?

As @jkotas suggested we'll use the PredefinedCulturesOnly config switch to control the behavior. I'll add more details in the PR description about that.

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:13]: @eerhardt could you please have a look at the change in this file and the change in the test .proj file too and confirm it is correct? thanks!

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:958]:
```suggestion
    <value>Only the invariant culture is supported in globalization-invariant mode. Either turn off globalization-invariant mode or use only the invariant culture. See https://aka.ms/GlobalizationInvariantMode for more information.</value>
```

I see the docs call it 'globalization-invariant mode'

File: src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs
Comment (danmoseley) [inline src/libraries/System.Globalization/tests/CultureInfo/GetCultureInfo.cs:156]: Nit, formatting of braces on these lines

**Review by danmoseley — APPROVED** [review]

**Comment by danmoseley** [comment]
(Signed off but someone else should check the linker/trimming stuff.)

File: src/libraries/System.Globalization/tests/Invariant/runtimeconfig.template.json
Comment (jkotas) [inline src/libraries/System.Globalization/tests/Invariant/runtimeconfig.template.json:4]: Should we rather test `PredefinedCulturesOnly: true` since it is the new default?

Comment (tarekgh) [reply src/libraries/System.Globalization/tests/Invariant/runtimeconfig.template.json:4]: We already testing it in the new test case added in the file GetCultureInfo.cs. I want to keep this Invariant test be exercised with the old behavior to ensure we are not regressing either way.

**Comment by tarekgh** [comment]
@lewing @qfyin I am seeing the failure https://github.com/dotnet/runtime/pull/54247/checks?check_run_id=2909713473. 
Does WASM runs somehow force enabling the config switch `System.Globalization.PredefinedCulturesOnly`?

**Comment by jkotas** [comment]
https://github.com/dotnet/runtime/blob/ea1707cf3a4a513de37a41d89591cca88460e9fa/src/mono/wasm/runtime/library_mono.js#L1336

**Comment by tarekgh** [comment]
Thanks @jkotas. Any idea if the app config switch is used in such runs? usually, app config switch takes precedence over the environment variable.

**Comment by jkotas** [comment]
This code suggests that DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY is unconditionally enabled for Wasm and there is no way to disable it.

**Comment by tarekgh** [comment]
> This code suggests that DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY is unconditionally enabled for Wasm and there is no way to disable it.

The Invariant test has a config switch which disable this predefined culture. The config switch should take higher precedence over DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY environment variable. So, my question is, does WASM runs not respecting the config switches on the tests? @lewing do you know how things work here?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:672]: Why not move it to the check below and remove the invariant assert there?

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:13]: Hmm, I think this is a bit problematic as it will need SDK update to pass the feature (System.Globalization.PredefinedCulturesOnly) value based on other feature (System.Globalization.Invariant) value

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs:1099]: It might be worth extracting into a simple method with single name argument instead of 4 same copies

**Comment by tarekgh** [comment]
> Hmm, I think this is a bit problematic as it will need SDK update to pass the feature (System.Globalization.PredefinedCulturesOnly) value based on other feature (System.Globalization.Invariant) value

@marek-safar @eerhardt could you please advise if I need to do anything here? should I hold this change? or SDK changes can be done after I merge this change? anything else in general I need to do?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs]: done moving it.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs]: I have done that.

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:13]: @marek-safar Marek Safar FTE @eerhardt Eric Erhardt FTE could you please advise if I need to do anything here? should I hold this change? or SDK changes can be done after I merge this change? anything else in general I need to do?

File: src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs
Comment (eerhardt) [inline src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs:17]: Did you intend for this property to call `PredefinedCulturesOnly()` every time it is invoked? Or did you want it to be a static field that is initialized once?

File: src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs
Comment (eerhardt) [inline src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs:511]: Are these tests ever invoked with PredefinedCulturesOnly not disabled? It is kind of odd for every test in this assembly to be marked conditional on it - why not just not run the tests in that mode?

File: src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs:18]: Can we find another way to test that Invariant mode is set here? This is basically removing what this test is trying to test.

Maybe a thought here is to split this test into 2?
* One just set Invariant = true, and runs this part of code.
* One sets both Invariant = true && PredefinedCultures = true, and tests that the right things are trimmed from GlobalizationMode.

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml
Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:13]: > or SDK changes can be done after I merge this change?

Yes, once we merge this change into runtime, we can make the change to the SDK

**Review by eerhardt — APPROVED** [review]
The changes look good to me. It would be great if we could keep the trimming test to still test the functionality somehow.

File: src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs
Comment (tarekgh) [reply src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs:511]: Yes, it was failing for all WASM runs. Look at the discussion in the comments https://github.com/dotnet/runtime/pull/54247#issuecomment-868093885 and the reply there. If you have a better idea to handle that I can do it.

File: src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs
Comment (tarekgh) [reply src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs:18]: I am not sure what this is providing more what we test in other tests. We have a complete test in the Invariant test project which is testing Invariant mode when PredefinedCultures is set to false. 

When having Invariant mode is on and PredefinedCultures is true, we'll not be able to create any culture tr-TR anyway and we'll throw exception. 

Could you please provide mode details what is we are getting to have tr-TR check here?

File: src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs
Comment (tarekgh) [reply src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs:17]: Although it is not a big deal to have this property called a few times, but it is not required to do so. I'll convert it to a static property.

File: src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs
Comment (eerhardt) [reply src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs:18]: > I am not sure what this check providing more than what we are doing in other Invariant tests.

No other tests are running against a trimmed app. This is the only test that sets `Invariant=true`, trims the app and then executes it to ensure it still works correctly.

> tr-TR

Using Turkish culture isn't the important thing here. The important thing is testing that Invariant mode is working correctly. We just happened to be using Turkish culture here as a way to test Invariant mode in a trimmed app.

See also the other side of this test, when Invariant=false:

https://github.com/dotnet/runtime/blob/daeee34715ea012d7a42af19137fd73f03b1f151/src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationFalse.cs#L13-L22

File: src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs
Comment (eerhardt) [reply src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs:511]: Ok, I guess that makes sense.

File: src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs
Comment (tarekgh) [reply src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs:18]: @eerhardt I have addressed this in my last commit.

File: src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs:30]: When is this ever false? It always gets replaced with true when the app is trimmed.

**Review by eerhardt — APPROVED** [review]
Just one question.

File: src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs
Comment (tarekgh) [reply src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs:30]: I was not sure if we run into any situation forcing it to false as we saw WASM was forcing it to true. anyway, if you are sure this will never be false then I'll modify the test according to that.

Comment (tarekgh) [reply src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs]: I have updated the test according to your assumption and let's see how it goes in CI.

Comment (eerhardt) [reply src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs]: > anyway, if you are sure this will never be false then I'll modify the test according to that.

You are explicitly passing `--feature System.Globalization.PredefinedCulturesOnly true` into the linker for this test. That tells the linker to replace `GlobalizationMode.PredefinedCulturesOnly`'s code to just return a hard-coded `true`. It can never be `false` in this test because of this.

File: docs/workflow/trimming/feature-switches.md
Comment (eerhardt) [inline docs/workflow/trimming/feature-switches.md:16]: `a made-up cultures` isn't grammatically correct.

How about 

```suggestion
| PredefinedCulturesOnly | System.Globalization.PredefinedCulturesOnly |  Don't allow creating a culture for which the platform does not have data |
```

?

Comment (tarekgh) [reply docs/workflow/trimming/feature-switches.md:16]: > Don't allow creating a culture for which the platform does not have data

feeling something missing. may be adding word `for` at the end make it more understandable.

**Comment by tarekgh** [comment]
System.Net failing tests are unrelated and tracked by the issue https://github.com/dotnet/runtime/issues/54778

**Comment by tarekgh** [comment]
@marek-safar @lewing @radical the WASM tests are failing (as expected) with this PR. This is happening because we are changing Invariant globalization mode to throw by default. Could you please advise how to fix these [failing tests](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-54247-merge-a6cc1fc3b6924bc185/Wasm.Build.Tests/console.98e0a037.log?sv=2019-07-07&se=2021-07-19T21%3A41%3A24Z&sr=c&sp=rl&sig=2bxJ%2BbPg%2BvYnQ8C0lpyrm8jD%2B%2FTYmqKOFfbvx%2BihHPY%3D)? note that we don't have msbuild property support yet. Is there a way I can define environment variable or config switch to use when building and running such tests?

**Comment by radical** [comment]
`RunAndTestWasmApp` takes an `envVars` argument also, you could set it in that. Will that be enough?

**Comment by tarekgh** [comment]
I tried setting the environment variable there, but it looks this didn't affect the test.

```
  [wasm test] info: Running v8 --expose_wasm runtime.js -- --setenv=DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY=false --setenv=XHARNESS_DISABLE_COLORED_OUTPUT=true --run invariant_True.dll
  [wasm test]
  [wasm test]
  [wasm test] fail: console.error: Error: System.Globalization.CultureNotFoundException: Only the invariant culture is supported in globalization-invariant mode. See https://aka.ms/GlobalizationInvariantMode for more information. (Parameter 'name')
  [wasm test]
  [wasm test] fail: es-ES is an invalid culture identifier.
  [wasm test]
```

The last idea I have is to force the value of PredefinedCulturesOnly in the test to false through the reflection. @marek-safar @lewing @radical if you have better idea please let me know.

**Comment by radical** [comment]
Can you share the change you made, and binlog for the test?

**Comment by tarekgh** [comment]
@radical the test change is just creating the envVars and passing it to `RunAndTestWasmApp`

```C$
            Dictionary<string, string> envVars = new();
            envVars.Add("DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY", "false");

            RunAndTestWasmApp(buildArgs, expectedExitCode: 42,
                                test: output => Assert.Contains(expectedOutputString, output), host: host, id: id, envVars: envVars);
```

Looking at the logs, I am seeing the test `src/tests/BuildWasmApps/Wasm.Build.Tests/InvariantGlobalizationTests.cs` will get trimmed during building. Looks the behavior would be defined during building and not during the runtime. if this true, it would be difficult to control that behavior for the newly introduced property `PredefinedCulturesOnly` as we don't have msbuild property controlling it, yet I guess.

I guess the best to do now, is to modify the test to catch the exception and use Invariant culture at that time. I'll try that and see how it goes.

/cc @eerhardt

**Comment by tarekgh** [comment]
I fixed the browser test, hopefully this will be last failure :-)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: Why are Invariant and PredefinedCulturesOnly not symmetric in their design?  e.g. I was expecting to see:
```C#
internal static bool Invariant { get; } = GetInvariantSwitchValue();
internal static bool PredefinedCulturesOnly { get; } = GetPredefinedCulturesOnlySwitchValue();
```

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: Because we decided to not introduce a new config switch with the new Invariant behavior. @jkotas suggested to use the `PredefinedCulturesOnly` switch which make a lot of sense. Now the default Invariant behavior is to throw for any non-Invariant culture. For that we have to have `PredefinedCulturesOnly` be true by default for Invariant. and by false by default for non-Invariant mode.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: I was asking a style question. GetPredefinedCulturesOnlySwitchValue() would return `AppContextConfigHelper.GetBooleanConfig("System.Globalization.PredefinedCulturesOnly", "DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY", Invariant);`, exactly what you're already initializing it to.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: @stephentoub I am going to merge this change soon. please let me know if you have any concern.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: I just don't see why the shape of these is different.  But it's not a blocker, it's just inconsistent.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: Sorry my reply crossed your reply. This is how it is done to help the linker for trimming purpose I guess. look at @marek-safar change https://github.com/dotnet/runtime/commit/bc27d49a2ed20029fd1d3630e21cd2b54ef75d54 which introduced that.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: I think this done this way to allow the linker to trim `GetInvariantSwitchValue` when Invariant mode is one. Having `GetPredefinedCulturesOnlySwitchValue` will require more work from the linker to trim it too. @eerhardt or @marek-safar should have definitive answer here.

**Comment by tarekgh** [comment]
I opened the issue https://github.com/dotnet/runtime/issues/55026 to track the unrelated JIT test failure in Android x64.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs
Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: > I think this done this way to allow the linker to trim GetInvariantSwitchValue when Invariant mode is one.

I don't believe this is the case. I think it is inconsistent because this code has been refactored multiple times, and no one has made it consistent.

> Having GetPredefinedCulturesOnlySwitchValue will require more work from the linker to trim it too. 

I don't believe this is true either. Alternatively, the opposite approach could be taken to make the code look consistent:

```C#
internal static bool Invariant { get; } = AppContextConfigHelper.GetBooleanConfig("System.Globalization.Invariant", "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
internal static bool PredefinedCulturesOnly { get; } = AppContextConfigHelper.GetBooleanConfig("System.Globalization.PredefinedCulturesOnly", "DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY", Invariant);
```

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:13]: I have created the SDK draft PR https://github.com/dotnet/sdk/pull/18672. and added the entry to the document https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md

File: src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:15]: I'll try to submit a PR to make this consistent.

**Comment by tarekgh** [comment]
@ericstj the description of this issue already referenced the breaking change issue https://github.com/dotnet/docs/issues/24849 and it is already merged to the docs https://docs.microsoft.com/en-us/dotnet/core/compatibility/globalization/6.0/culture-creation-invariant-mode

**Pull request merged**
